### ISSUE TITLE
copy module does not copy when destination is a directory

### ISSUE THREAD

**Issue by srgvg** [ansible/ansible#2385]
on recent ansible, commit from March 13 2013

serge@goldorak:~$ ansible -i localhost, all -m copy -a "src=~/tmp/kaka dest=/var/tmp/"
localhost | success >> {
    "changed": false, 
    "dest": "/var/tmp/", 
    "group": "root", 
    "mode": "01777", 
    "msg": "file already exists", 
    "owner": "root", 
    "size": 4096, 
    "src": "/home/serge/.ansible/tmp/ansible-1363190397.35-165795465175179/kaka", 
    "state": "directory"
}

serge@goldorak:~$ ansible -i localhost, all -m copy -a "src=~/tmp/kaka dest=/var/tmp"
localhost | success >> {
    "changed": false, 
    "dest": "/var/tmp", 
    "group": "root", 
    "mode": "01777", 
    "msg": "file already exists", 
    "owner": "root", 
    "size": 4096, 
    "src": "/home/serge/.ansible/tmp/ansible-1363190406.58-141330975620440/kaka", 
    "state": "directory"
}

**Comment by mpdehaan** [comment]
In the last example, you do need to end in a trailing slash, though I will look into the first.

**Comment by britcey** [comment]
I thought it should handle it even without a trailing slash?

Regardless, in copy.py on line 126 it calls the module with the provided arguments:

```
        module_args = "%s src=%s" % (module_args, pipes.quote(tmp_src))
```

but back on line 99 it already figured out that dest was pointing to a directory:

```
        dest = os.path.join(dest, os.path.basename(source))
```

but 'module_args' never picks up the updated dest, so the call to the copy module has bad info.

copy also has some (broken?) logic to deal with directories as destination, but since copy.py already knows the destination is a directory it should call copy with the dest already updated to reflect that fact.

**Comment by britcey** [comment]
Not sure if this is related, but 'force' is now defaulting to 'false' for copy - I have to suspect the recent changes to argument checking with 'file' - let me know if you'd like this pulled out to a separate issue.

**Comment by mpdehaan** [comment]
I just removed some code that implemented the force (default True, optional False) option to copy badly that was causing problems here.  I suspect it will work fine now, though fixing "force" on copy will need to be done again.

**Comment by srgvg** [comment]
Looks like this bug also affects the template module, but as this is based on the same file module I expect this to be related.

The file does not get updated when it already exists:

``` yaml
root@ansible-pr-1:~/tmp# cat testtemplate.yml 
---
- name: test template
  hosts: all
  gather_facts: false
  tasks:
  - local_action: template src=~/tmp/kaka dest=~/tmp/kaka2
root@ansible-pr-1:~/tmp# cat kaka
{{ ansible_managed }}
root@ansible-pr-1:~/tmp# cat kaka2
cat: kaka2: No such file or directory

root@ansible-pr-1:~/tmp# ansible-playbook -i localhost, testtemplate.yml -v

PLAY [test template] ********************* 

TASK: [template src=~/tmp/kaka dest=~/tmp/kaka2] ********************* 
ok: [localhost] => {"changed": false, "dest": "/root/tmp/kaka2", "group": "root", "mode": "0644", "msg": "file already exists", "owner": "root", "size": 88, "src": "/root/.ansible/tmp/ansible-1363256756.77-240540220307896/source", "state": "file"}

PLAY RECAP ********************* 
localhost                      : ok=1    changed=0    unreachable=0    failed=0    




root@ansible-pr-1:~/tmp# cat kaka2
Ansible managed: /root/tmp/kaka modified on 2013-03-14 11:20:22 by root on ansible-pr-1
root@ansible-pr-1:~/tmp# vi kaka
root@ansible-pr-1:~/tmp# cat kaka
{{ ansible_managed }}
{{ ansible_managed }}
root@ansible-pr-1:~/tmp# ansible-playbook -i localhost, testtemplate.yml 

PLAY [test template] ********************* 

TASK: [template src=~/tmp/kaka dest=~/tmp/kaka2] ********************* 
ok: [localhost]

PLAY RECAP ********************* 
localhost                      : ok=1    changed=0    unreachable=0    failed=0    


root@ansible-pr-1:~/tmp# cat kaka2
Ansible managed: /root/tmp/kaka modified on 2013-03-14 11:20:22 by root on ansible-pr-1
root@ansible-pr-1:~/tmp# 

```

**Comment by dhozac** [comment]
Should be fixed by #2395.

**Comment by srgvg** [comment]
OK, I rechecked all my tests, and this all works no, including
- mentioning a filename in the destination
- mentioning a destination directory only, with a trailing /
- mentioning a destination directory only, without a trailing  /

Same for the template module, though there is 1 edge case performing differently:

``` yaml
TASK: [template src=~/tmp/kaka dest=/var/tmp] ********************* 
changed: [localhost] => {"changed": true, "dest": "/var/tmp/source", "group": "serge", "md5sum": "a117449685eecc05b80965c3d8f59a14", "mode": "0664", "owner": "serge", "size": 182, "src": "/home/serge/.ansible/tmp/ansible-1363278732.25-53084373351816/source", "state": "file"}
```

vs 

``` yaml
TASK: [template src=~/tmp/kaka dest=/var/tmp/] ********************* 
changed: [localhost] => {"changed": true, "dest": "/var/tmp/kaka", "group": "serge", "md5sum": "a117449685eecc05b80965c3d8f59a14", "mode": "0664", "owner": "serge", "size": 182, "src": "/home/serge/.ansible/tmp/ansible-1363278758.27-176582105796064/source", "state": "file"}
```

so when the destination is a directory, and not mentioned with a trailing /, the destination filename becomes _source_.

**Comment by mpdehaan** [comment]
Seems like you're running an old version of ansible or something, checkout or otherwise

**Comment by srgvg** [comment]
I'm afraid it that bug still is current:

``` bash

serge@cyberlab:~/tmp$ cat testtemplate.yml 
---
- name: test template
  hosts: all
  gather_facts: false
  tasks:
  - local_action: template src=~/tmp/kaka dest=/var/tmp
serge@cyberlab:~/tmp$ ansible --version
ansible 1.1

serge@cyberlab:~/tmp$ (cd ../src/ansible/; git pull; make deb >/dev/null 2>&1 ; sudo dpkg -i ../ansible_1.1_all.deb)
Current branch devel is up to date.

(Reading database ... 470851 files and directories currently installed.)
Preparing to replace ansible 1.1 (using ../ansible_1.1_all.deb) ...
Unpacking replacement ansible ...
Setting up ansible (1.1) ...
Processing triggers for python-support ...
Processing triggers for man-db ...

serge@cyberlab:~/tmp$ ansible-playbook -i localhost, testtemplate.yml -v

PLAY [test template] ********************* 

TASK: [template src=~/tmp/kaka dest=/var/tmp] ********************* 
ok: [localhost] => {"changed": false, "dest": "/var/tmp/source", "group": "serge", "md5sum": "a117449685eecc05b80965c3d8f59a14", "mode": "0664", "owner": "serge", "size": 182, "src": "/home/serge/.ansible/tmp/ansible-1364066410.03-276763480209852/source", "state": "file"}

PLAY RECAP ********************* 
localhost                      : ok=1    changed=0    unreachable=0    failed=0    



serge@cyberlab:~/tmp$ vi testtemplate.yml
serge@cyberlab:~/tmp$ cat testtemplate.yml 
---
- name: test template
  hosts: all
  gather_facts: false
  tasks:
  - local_action: template src=~/tmp/kaka dest=/var/tmp/



serge@cyberlab:~/tmp$ ansible-playbook -i localhost, testtemplate.yml -v

PLAY [test template] ********************* 

TASK: [template src=~/tmp/kaka dest=/var/tmp/] ********************* 
ok: [localhost] => {"changed": false, "group": "serge", "mode": "0664", "owner": "serge", "path": "/var/tmp/kaka", "size": 182, "state": "file"}

PLAY RECAP ********************* 
localhost                      : ok=1    changed=0    unreachable=0    failed=0 

```

**Comment by mpdehaan** [comment]
testing this now...

**Comment by mpdehaan** [comment]
Fixed in 3c5890f ... I've tested with both template and copy.

Additional testing quite welcome, this seems solid though.

It's a rather weird trick where we pass in a hidden argument to the copy module that tells it what the original filename was.  I did not like the idea of using the actual filename because it might clash with the module name in very rare instances.

Let me know if this works for you.  Thanks!
